Table Visualization for TSVB

simianhacker opened the pull request:
This PR adds a table visualization for Time Series Visual Builder. For the most part everything works like the rest of the visualizations with a few exceptions:

- The user MUST pick a `group by field` to aggregate the data on.
- Series Aggs are not available via the normal aggregation picker. The same functionality is located in the series options. 
- Series can have filters that are applied only to that series (or column). This is useful for things like network traffic where you want to aggregate the series by `system.network.name` but you want to filter using `-system.network.name:lo*` to remove localhost.
- Time shifting is removed (since everything is run with the same aggregation)
- Sorting works only on columns where the last metric is a basic aggregation (or sortable)

*Editor*

![image](https://user-images.githubusercontent.com/41702/28173054-2232333a-67a2-11e7-8173-3632f24f3c80.png)


*On a dashboard*

![image](https://user-images.githubusercontent.com/41702/28173106-49e5b4a6-67a2-11e7-96ca-023ec63d473e.png)


*Series Options*

![image](https://user-images.githubusercontent.com/41702/28140342-d204156c-670c-11e7-9a0b-36bc7f9d81e0.png)

*Panel Options*

![image](https://user-images.githubusercontent.com/41702/28173075-36108abe-67a2-11e7-9898-1a5d20c4c07e.png)

trevan commented:
@simianhacker, what do you mean by "id_field to pivot data on"?

I'm also curious why you are using different sort icons from the rest of the app?

simianhacker commented:
@trevan pivot is probably an overloaded term, I changed it to "group by" which is really just the field to run a terms aggregation on at the top of the aggregation tree. The sort icons match what's in monitoring, I just noticed that it's different the the rest of the application. I updated to what is standard in the other Kibana visualizations. (Thanks for your feedback :D)

simianhacker commented:
Jenkins test this

monicasarbu commented:
@simianhacker Would it be possible to add a search box near the `Hosts` column to filter for a particular host in the table? I think this would be quicker than trying to use the overall Kibana search box.

timroes commented on src/core_plugins/metrics/public/components/panel_config/table.js:
```diff
@@ -0,0 +1,150 @@
+            <div className="vis_editor__label">Item Url (This supports mustache templating.
+              <code>{'{{key}}'}</code> is set to the term)
+            </div>
+            <input
```
This input is missing a [correct label](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element).

timroes commented on src/core_plugins/metrics/public/components/panel_config/table.js:
```diff
@@ -0,0 +1,150 @@
+          />
+          <div className="vis_editor__vis_config-row">
+            <div className="vis_editor__label">Panel Filter</div>
+            <input
```
This input is missing a [correct label](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element).

timroes commented on src/core_plugins/metrics/public/components/panel_config/table.js:
```diff
@@ -0,0 +1,150 @@
+            onChange={this.props.onChange}
+          />
+          <div className="vis_editor__vis_config-row">
+            <div className="vis_editor__label">Panel Filter</div>
```
This should most likely be a `<label>` with a `for` attribute pointing to the `input`.

timroes commented on src/core_plugins/metrics/public/components/vis_types/table/config.js:
```diff
@@ -0,0 +1,113 @@
+              value={model.formatter}
+            />
+            <div className="vis_editor__label">Template (eg.<code>{'{{value}}/s'}</code>)</div>
+            <input
```
This input is missing a [correct label](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element).

timroes commented on src/core_plugins/metrics/public/components/vis_types/table/config.js:
```diff
@@ -0,0 +1,113 @@
+          </div>
+          <div className="vis_editor__series_config-row">
+            <div className="vis_editor__label">Filter</div>
+            <input
```
This input is missing a [correct label](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element).

timroes commented on src/core_plugins/metrics/public/components/vis_types/table/config.js:
```diff
@@ -0,0 +1,113 @@
+            </div>
+            <div className="vis_editor__label">Aggregate Function</div>
+            <div className="vis_editor__row_item">
+              <Select
```
This element is missing a [proper label](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element).

You can add an `id` to the actual input field in `react-select` using:

```js
inputProps={{ id: 'someid' }}
```

timroes commented on src/core_plugins/metrics/public/components/vis_types/table/series.js:
```diff
@@ -0,0 +1,155 @@
+        <div className="vis_editor__series-details">
+          <div onClick={props.toggleVisible}><i className={caretClassName}/></div>
+          <div className="vis_editor__row vis_editor__row_item">
+            <input
```
This input is missing a [correct label](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#label-every-form-element).

timroes commented on src/core_plugins/metrics/public/components/vis_types/table/series.js:
```diff
@@ -0,0 +1,155 @@
+    >
+      <div className="vis_editor__container">
+        <div className="vis_editor__series-details">
+          <div onClick={props.toggleVisible}><i className={caretClassName}/></div>
```
If this is clickable it should propably be a `button`. https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#use-button-and-a-href

timroes commented on src/core_plugins/metrics/public/components/vis_types/table/series.js:
```diff
@@ -100,9 +100,14 @@ function TopNSeries(props) {
         <div className="vis_editor__series-details">
-          <div onClick={props.toggleVisible}><i className={caretClassName}/></div>
+          <div>
+            <a onClick={props.toggleVisible}>
```
This is an action and not a navigation. Could you use a [`button` instead of an `a`](https://github.com/elastic/kibana/blob/master/style_guides/accessibility_guide.md#use-button-and-a-href) for this, please, since an `a` without an `href` isn't different from a `div` for a screen reader and also semantics and stuff :-)

simianhacker commented:
jenkins test this

simianhacker commented:
Jenkins test this

timroes commented on src/core_plugins/metrics/public/components/panel_config/table.js:
```diff
@@ -0,0 +1,155 @@
+                </p>
+              </div>
+              <div className="vis_editor__vis_config-row">
+                <label className="vis_editor__label">Group By Field</label>
```
I've just merged the a11y TSVB PR. Could you maybe rebase to master and also label this, the `FieldSelect` now supports an `id` property, and there is the `htmlIdGenerator` to generate the ids. (I skip a long explanation here, since you reviews that PR anyway :D)

shaharmor commented:
How can I test this locally with v6.0.0-beta2 of ES?

timroes commented:
Sorry for bothering, but I just moved the `html_id_generator` outside of the plugin (PR #13906), since other plugins were beginning to require it also. Could you please rebase this PR to `master` and change the import to `from 'ui_framework/services'` - everything else will stay the same. I backported that to 6.x and 6.0, so your PR should work fine when backporting, though I would appreciate, some extra attention when cherry-picking - and I owe you a :beer:!

simianhacker commented:
@shaharmor The easiest way to test this PR right now would be to setup a development environment or wait till we merge and backport this for you :D (Hopefully it will only be a few days)

shaharmor commented:
@simianhacker I can't wait :) This TSVB is probably the best thing that happened to Kibana in a very long time. Really awesome.

simianhacker commented:
Jenkins test this

simianhacker commented:
@shaharmor Wow... thanks! That made my week and it's only Monday!

shaharmor commented:
@simianhacker
When using the table to show only non-basic-aggregation fields, the results that show up may be incorrect.

The reason is because there is no sort option that applies to all series, which makes it that each series will choose the top X results based on its own metric.
This may cause other series to be half full (Have `null` values for some buckets) and show incorrect data.

simianhacker commented:
jenkins test this

simianhacker commented:
@shaharmor When there isn't a column to sort by (because the last metric isn't a sortable aggregation) it uses `_terms` to sort on `asc`.

shaharmor commented:
So if I want to display the top X terms by cumulative sum of some field I'm out of luck?
Because I'll have no basic aggregation, and the _terms sort will leave me with the first X terms, not the top ones

simianhacker commented:
@shaharmor Correct, it's a limitation of Elasticsearch's aggregation framework. You can't sort a terms aggregation on a pipeline or sibling aggregation. One solution would be to setup a script that runs every minutes or so that aggregates your data into a new index then run simple aggs on those doc's for sorting and what not.

simianhacker commented:
@thomasneirynck Any chance we could get this reviewed soon? We (Beats) would like to start using this in the OSS Metricbeat dashboards.

shaharmor commented:
@simianhacker 
1. Is it possible to add a client-side sort? Once the table exists, it should be pretty easy to add sort to all columns?
2. Can you add pagination? Like in the normal table visualization?

shaharmor commented:
When using a bucket script aggregation and some of the data is missing (returned as `null` from the server), the entire result is 0

shaharmor commented:
So, following the understanding the pipelines aggregations are not sortable, I would like to suggest another feature for the table visualization (And really for all not-time-based visualizations in TSVB).

I understand that "time-series" is the whole point of TSVB, but it also stops us from having some required features.

I'll elaborate:
We have docs that look like this:
```json
{
  "product": "computer",
  "price": 100,
  "count": 5
}
```

We want to have a table with 3 columns:
1. the name of the product (The "dimension")
2. The sum of the prices of all the docs (The "metric")
3. The ratio between the sum of the prices and the sum of counts (The "calculation").
And have the table be sorted by the sum of prices.

Because TSVB always add a `date_histogram` aggregation as the first aggregation, in order to get the sum of the prices, we need to use 2 aggregations:
1. sum of price field
2. cumulative sum the "sum of price field"

This means that we have to use a pipeline aggregation, and then lose sorting.

If we remove the `date_histogram` aggregation, we could just use the "sum of price field" to get the same results, plus making the column sortable.

I suggest to add an option to remove the `date_histogram` aggregation on a per-visualization basis (Basically in the "Panel Options").

simianhacker commented:
@shaharmor I think you should make a feature request issue for adding calculation to core Kibana's data table visualization. Since that one doesn't have a date_histogram by default and will do everything you want with the exception of the calculation.

shaharmor commented:
I will definitely do that as well, although I really think TSVB can also
gain from a feature like that, both in terms of the visualization and in
terms of query performance (why do an aggregation that's not used?)

On Thu, 28 Sep 2017 at 19:40 Chris Cowan <notifications@github.com> wrote:

> @shaharmor <https://github.com/shaharmor> I think you should make a
> feature request issue for adding calculation to core Kibana's data table
> visualization. Since that one doesn't have a date_histogram by default and
> will do everything you want with the exception of the calculation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/kibana/pull/12813#issuecomment-332894091>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKW9YEhbYDXyXAhLG9q7urbSa8NgBuwSks5sm8wGgaJpZM4OWNVy>
> .
>

simianhacker commented:
@shaharmor The `date_histogram` is being used and is at the heart of TSVB, even in the TopN, Gauge, Metric and Table visualizations. This is what allows users to create metrics like "current network throughput" or any kind of rate. You need the historical data to do the derivative for those metrics. Also currently in Kibana there isn't any other way to show the value of the "last bucket" or what could be considered the "near real time" value.

mattapperson commented:
When creating a a new vis and selecting table I am getting the error of
```
TypeError: Cannot read property 'order' of undefined
    at doc (/Users/mattapperson/Development/kibana/src/core_plugins/metrics/server/lib/vis_data/request_processors/table/pivot.js:47:51
```

shaharmor commented:
When is this going to be merged? Are you waiting for the 6.0 GA in order to merge it to the 6.x branch?

simianhacker commented:
@shaharmor Probably some time today :D

simianhacker commented:
jenkins test this

simianhacker commented:
jenkins test this

simianhacker commented:
Back ported to 6.x with fe7e8a59d

shaharmor commented on src/core_plugins/metrics/server/lib/vis_data/request_processors/table/date_histogram.js:
```diff
@@ -0,0 +1,26 @@
+        min_doc_count: 0,
+        extended_bounds: {
+          min: from.valueOf(),
+          max: to.valueOf() - (bucketSize * 1000)
```
@simianhacker why do you reduce the `bucketSize` here?

simianhacker replied:
This was removed from the rest of TSVB but because this PR languished in review for so long I forgot to remove it from here as well. I have a PR that fixes this. #14553

shaharmor replied:
Thnx, should probably be removed from here as well no? https://github.com/elastic/kibana/pull/12813/files/abeeaa7224fd391aa73e3b1200888da3623e725c#diff-63d4f38a0a2ce1a25e2e18de8982c1b1R21

simianhacker replied:
Oops... Added that one as well to the PR... Thanks!

shaharmor commented on src/core_plugins/metrics/server/lib/vis_data/request_processors/table/pivot.js:
```diff
@@ -0,0 +1,53 @@
+              filter: {
+                range: {
+                  [timeField]: {
+                    gte: to.valueOf() - (bucketSize * 1500),
```
Also here, why change the value?

simianhacker replied:
This is here for sorting, the group by terms works the same way (https://github.com/elastic/kibana/blob/master/src/core_plugins/metrics/server/lib/vis_data/request_processors/series/split_by_terms.js#L26-L44). You can't sort a terms aggregation on a metric that is a child of a date histogram so you have to create a metric outside the date histogram to sort on with a filter that approximates the last bucket. I originally took the bucket size and multiplied by 1000 milliseconds but would occasionally get strange results; using 1500 milliseconds made things more reliable. Also notice this is slightly different because it's adding the time to the `gte` side and not the `lte` side.

Shifter2600 commented:
Issue when using a Date field as the Group By Field.
I have a field that is formatted to YYYY-MM and when I try to use that as my group by I get
1443657600000
1427846400000
1430438400000
....
The Columns should be taking the Field Specific Format
